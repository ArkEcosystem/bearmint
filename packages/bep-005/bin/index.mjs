#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
