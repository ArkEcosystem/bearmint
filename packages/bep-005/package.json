{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
