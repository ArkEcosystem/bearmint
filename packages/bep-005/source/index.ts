export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
