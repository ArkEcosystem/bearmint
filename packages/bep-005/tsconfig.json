{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
