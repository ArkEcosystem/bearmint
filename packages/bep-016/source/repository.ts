import { assert } from '@bearmint/bep-009'
import type { Cradle, ServiceProvider, ServiceProviderRepository } from '@bearmint/bep-013'
import { Event } from '@bearmint/bep-013'

export function makeServiceProviderRepository(
	cradle: Pick<Cradle, 'Container' | 'EventDispatcher'>,
): ServiceProviderRepository {
	const serviceProviders = new Map<string, ServiceProvider>()
	const loadedProviders = new Set<string>()
	const failedProviders = new Set<string>()
	const deferredProviders = new Set<string>()

	function get(name: string): ServiceProvider {
		const serviceProvider: ServiceProvider | undefined = serviceProviders.get(name)

		assert.defined<ServiceProvider>(serviceProvider)

		return serviceProvider
	}

	return {
		all() {
			return serviceProviders
		},
		allLoadedProviders() {
			return [...loadedProviders.values()].map((name: string) => get(name))
		},
		async boot(name: string) {
			await cradle.EventDispatcher.dispatch(Event.ServiceProviderBooting, {
				name,
			})

			await get(name).boot()

			loadedProviders.add(name)
			failedProviders.delete(name)
			deferredProviders.delete(name)

			await cradle.EventDispatcher.dispatch(Event.ServiceProviderBooted, {
				name,
			})
		},
		defer(name: string) {
			deferredProviders.add(name)
		},
		deferred(name: string) {
			return deferredProviders.has(name)
		},
		async dispose(name: string) {
			await cradle.EventDispatcher.dispatch(Event.ServiceProviderDisposing, {
				name,
			})

			await get(name).dispose()

			loadedProviders.delete(name)
			failedProviders.delete(name)
			deferredProviders.add(name)

			await cradle.EventDispatcher.dispatch(Event.ServiceProviderDisposed, {
				name,
			})
		},
		fail(name: string) {
			failedProviders.add(name)
		},
		failed(name: string) {
			return failedProviders.has(name)
		},
		get,
		has(name: string) {
			return serviceProviders.has(name)
		},
		load(name: string) {
			loadedProviders.add(name)
		},
		loaded(name: string) {
			return loadedProviders.has(name)
		},
		async register(name: string) {
			await cradle.EventDispatcher.dispatch(Event.ServiceProviderRegistering, {
				name,
			})

			cradle.Container.bindValue(name, get(name).config())

			await get(name).register()

			await cradle.EventDispatcher.dispatch(Event.ServiceProviderRegistered, {
				name,
			})
		},
		set(name: string, provider: ServiceProvider) {
			serviceProviders.set(name, provider)
		},
	}
}
