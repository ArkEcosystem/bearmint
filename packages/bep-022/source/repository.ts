import { assert, bytesToHex, bytesToString, hexToBytes, stringToBytes } from '@bearmint/bep-009'
import type {
	Account,
	AccountRepositoryFactory,
	AccountWithMoniker,
	AccountWithPublicKey,
	AccountWithValidator,
	Cradle,
	KVStore,
} from '@bearmint/bep-013'
import { Event, PROTO_OPTS } from '@bearmint/bep-013'
import { BEP22Account } from '@bearmint/bep-018'

export function makeAccountRepository(cradle: Cradle): AccountRepositoryFactory {
	return {
		make(tries) {
			async function findByAddress(address: string) {
				const result = await tries.account.get(address)

				assert.defined<Buffer>(result)

				return cradle.AccountSerializer.deserialize(BEP22Account.fromBinary(result, PROTO_OPTS))
			}

			async function findByProxy<T>(trie: KVStore, key: string) {
				const address = await trie.get(key)

				assert.defined<Buffer>(address)

				return (await findByAddress(bytesToString(address))) as unknown as T
			}

			return {
				async allMonikers() {
					return (await tries.accountMoniker.keys()).map((moniker) => bytesToString(moniker))
				},
				async allValidatorAddresses() {
					return (await tries.accountValidatorAddress.keys()).map((address) => bytesToHex(address))
				},
				findByAddress,
				async findByMoniker<T = AccountWithMoniker>(moniker: string) {
					return findByProxy<T>(tries.accountMoniker, moniker)
				},
				async findByPublicKey(publicKey: string) {
					return findByProxy<AccountWithPublicKey>(tries.accountPublicKey, publicKey)
				},
				async findByValidatorAddress(address: string) {
					return findByProxy<AccountWithValidator>(tries.accountValidatorAddress, address)
				},
				async forgetByAddress(address: string) {
					await tries.account.forget(address)
				},
				async forgetByMoniker(moniker: string) {
					return tries.accountMoniker.forget(moniker)
				},
				async forgetByPublicKey(publicKey: string) {
					return tries.accountPublicKey.forget(publicKey)
				},
				async forgetByValidatorAddress(address: string) {
					return tries.accountValidatorAddress.forget(address)
				},
				async hasByAddress(address: string) {
					return tries.account.has(address)
				},
				async hasByMonkiker(moniker: string) {
					return tries.accountMoniker.has(moniker)
				},
				async hasByPublicKey(publicKey: string) {
					return tries.accountPublicKey.has(publicKey)
				},
				async hasByValidatorAddress(address: string) {
					return tries.accountValidatorAddress.has(address)
				},
				async index(items: Account[]) {
					for (const account of items) {
						if (account.address !== undefined) {
							await tries.account.set(
								account.address,
								(await cradle.AccountSerializer.serialize(account)).toBinary(),
							)
						}

						if (account.publicKey !== undefined) {
							await tries.accountPublicKey.set(
								hexToBytes(account.publicKey),
								stringToBytes(account.address),
							)

							if (account.moniker !== undefined) {
								await tries.accountMoniker.set(
									stringToBytes(account.moniker),
									stringToBytes(account.address),
								)
							}

							if (account.validator?.address !== undefined) {
								await tries.accountValidatorAddress.set(
									hexToBytes(account.validator.address),
									stringToBytes(account.address),
								)
							}
						}

						await cradle.EventDispatcher.dispatch(Event.AccountIndexed, account)
					}
				},
			}
		},
	}
}
