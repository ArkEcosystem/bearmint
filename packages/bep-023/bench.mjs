import { bench } from '@bearmint/bep-004'

import { makeSecureKVStore } from './distribution/index.js'

const trie = await makeSecureKVStore().make()

await bench('Merkle Patricia Trie', async ({ mark }) => {
	await mark('set -> get', async () => {
		await trie.set('key', Buffer.from('value'))
		await trie.get('key')
	})

	await mark('forget', async () => {
		await trie.forget('key')
	})

	await mark('has', async () => {
		await trie.has('key')
	})
})
