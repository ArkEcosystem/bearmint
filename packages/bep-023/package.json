{
    "name": "@bearmint/bep-023",
    "version": "0.0.0",
    "description": "This library is the official reference implementation for BEP-023 (https://bep.dev/023)",
    "keywords": [
        "bearmint",
        "tendermint",
        "blockchain",
        "framework",
        "modular"
    ],
    "homepage": "https://github.com/bearmint/bearmint",
    "repository": {
        "type": "git",
        "url": "git+https://github.com/bearmint/bearmint.git"
    },
    "funding": "https://funding.bearmint.com/",
    "license": "MIT",
    "author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
    "type": "module",
    "exports": "./distribution/index.js",
    "types": "./distribution/index.d.ts",
    "files": [
        "distribution"
    ],
    "scripts": {
        "audit": "pnpm audit --fix --prod",
        "bench": "node bench.mjs",
        "build": "pnpm run clean && tsc",
        "build:watch": "pnpm run clean && tsc -w",
        "clean": "rm -rf distribution",
        "deps:major": "ncu -u --target=latest",
        "deps:minor": "ncu -u --target=minor",
        "deps:patch": "ncu -u --target=patch",
        "docs": "typedoc source",
        "knip": "knip",
        "lint": "eslint 'source/**/*.ts' --format=codeframe --config=../../config/eslintrc.js --fix",
        "test": "pnpm run test:unit",
        "test:coverage": "pnpm run test:unit:coverage",
        "test:ui": "vitest --ui",
        "test:unit": "vitest run tests/unit",
        "test:unit:coverage": "vitest run tests/unit --coverage",
        "test:unit:ui": "vitest tests/unit --ui",
        "test:unit:watch": "vitest tests/unit",
        "test:watch": "vitest"
    },
    "dependencies": {
        "@bearmint/bep-009": "workspace:^",
        "@bearmint/bep-013": "workspace:^",
        "@bearmint/bep-016": "workspace:^",
        "@bearmint/bep-109": "workspace:^",
        "@ethereumjs/trie": "5.0.1",
        "classic-level": "1.2.0",
        "memory-level": "1.0.0"
    },
    "devDependencies": {
        "@bearmint/bep-004": "workspace:^",
        "@bearmint/bep-005": "workspace:^",
        "@types/readable-stream": "2.3.14"
    },
    "engines": {
        "node": ">=18",
        "pnpm": ">=7.9.0"
    },
    "publishConfig": {
        "access": "public"
    }
}
