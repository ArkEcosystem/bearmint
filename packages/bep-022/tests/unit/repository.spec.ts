import { expect, describe, it, beforeEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import type { AccountRepository } from '@bearmint/bep-013'
import { makeSecureKVStore } from '@bearmint/bep-023'

import {
	fakeAccountSerializer,
	fakeAddressFactory,
	fakeEventDispatcher,
	fakeMultiStore,
} from '../../../bep-008/distribution/index.js'
import { makeAccountRepository } from '../../source/repository.js'

const address = 'bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr'
const moniker = 'moniker'
const publicKey = '134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e'
const validator = {
	address: '2370a54078de4a12becff8746e1c2826c9ef8b92',
	delegators: { a: '0' },
	power: '0',
	publicKey: 'publicKey',
}

describe<{
	subject: AccountRepository
}>('Repository', () => {
	beforeEach(async (context) => {
		const multiStore = await fakeMultiStore()

		context.subject = makeAccountRepository({
			AccountSerializer: fakeAccountSerializer(),
			AddressFactory: fakeAddressFactory(),
			EventDispatcher: fakeEventDispatcher(),
		}).make({
			account: multiStore.get('@bearmint/bep-020/account'),
			accountMoniker: multiStore.get('@bearmint/bep-020/account_moniker'),
			accountPublicKey: multiStore.get('@bearmint/bep-020/account_public_key'),
			accountValidatorAddress: multiStore.get('@bearmint/bep-020/account_validator_address'),
		})
	})

	it('should get all monikers', async (context) => {
		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.allMonikers()).toHaveLength(1)
	})

	it('should get all validator addresses', async (context) => {
		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.allValidatorAddresses()).toHaveLength(1)
	})

	it('should index an account and find it by its address', async (context) => {
		expect(await context.subject.hasByAddress(address)).toBeFalse()

		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.hasByAddress(address)).toBeTrue()
		expect(await context.subject.findByAddress(address)).toMatchSnapshot()
	})

	it('should index an account and find it by its public key', async (context) => {
		expect(await context.subject.hasByPublicKey(publicKey)).toBeFalse()

		await context.subject.index([
			{
				address,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.hasByPublicKey(publicKey)).toBeTrue()
		expect(await context.subject.findByPublicKey(publicKey)).toMatchSnapshot()
	})

	it('should index an account twice and find it by its public key', async (context) => {
		expect(await context.subject.hasByPublicKey(publicKey)).toBeFalse()

		await context.subject.index([
			{
				address,
				publicKey,
			},
		])

		await context.subject.index([
			{
				address,
				publicKey,
			},
		])

		expect(await context.subject.hasByPublicKey(publicKey)).toBeTrue()
	})

	it('should index an account and find it by its validator moniker', async (context) => {
		expect(await context.subject.hasByMonkiker(moniker)).toBeFalse()

		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.hasByMonkiker(moniker)).toBeTrue()
		expect(await context.subject.findByMoniker('moniker')).toMatchSnapshot()
	})

	it('should index an account and find it by its validator address', async (context) => {
		expect(await context.subject.hasByValidatorAddress(validator.address)).toBeFalse()

		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.hasByValidatorAddress(validator.address)).toBeTrue()
		expect(await context.subject.findByValidatorAddress(validator.address)).toMatchSnapshot()
	})

	it('should index an account and forget it by its address', async (context) => {
		expect(await context.subject.hasByAddress(address)).toBeFalse()

		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.hasByAddress(address)).toBeTrue()

		await context.subject.forgetByAddress(address)

		expect(await context.subject.hasByAddress(address)).toBeFalse()
	})

	it('should index an account and forget it by its public key', async (context) => {
		expect(await context.subject.hasByPublicKey(publicKey)).toBeFalse()

		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.hasByPublicKey(publicKey)).toBeTrue()

		await context.subject.forgetByPublicKey(publicKey)

		expect(await context.subject.hasByPublicKey(publicKey)).toBeFalse()
	})

	it('should index an account and forget it by its validator moniker', async (context) => {
		expect(await context.subject.hasByMonkiker(moniker)).toBeFalse()

		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.hasByMonkiker(moniker)).toBeTrue()

		await context.subject.forgetByMoniker(moniker)

		expect(await context.subject.hasByMonkiker(moniker)).toBeFalse()
	})

	it('should index an account and forget it by its validator address', async (context) => {
		expect(await context.subject.hasByValidatorAddress(validator.address)).toBeFalse()

		await context.subject.index([
			{
				address,
				moniker,
				publicKey,
				validator,
			},
		])

		expect(await context.subject.hasByValidatorAddress(validator.address)).toBeTrue()

		await context.subject.forgetByValidatorAddress(validator.address)

		expect(await context.subject.hasByValidatorAddress(validator.address)).toBeFalse()
	})
})
