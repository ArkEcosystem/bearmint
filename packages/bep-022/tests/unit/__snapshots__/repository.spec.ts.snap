// Vitest Snapshot v1

exports[`Repository > should index an account and find it by its address 1`] = `
{
  "address": "bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr",
  "balances": {
    "BEAR": 0n,
    "FIRE": 0n,
    "GOLD": 0n,
    "SEED": 0n,
  },
  "lockedBalances": {
    "BEAR": 0n,
    "FIRE": 0n,
    "GOLD": 0n,
    "SEED": 0n,
  },
  "metadata": {},
  "moniker": "moniker",
  "nonce": 0n,
  "publicKey": "134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e",
  "stakes": {},
  "validator": {
    "address": "2370a54078de4a12becff8746e1c2826c9ef8b92",
    "power": 0n,
    "publicKey": "publicKey",
  },
}
`;

exports[`Repository > should index an account and find it by its public key 1`] = `
{
  "address": "bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr",
  "balances": {
    "BEAR": 0n,
    "FIRE": 0n,
    "GOLD": 0n,
    "SEED": 0n,
  },
  "lockedBalances": {
    "BEAR": 0n,
    "FIRE": 0n,
    "GOLD": 0n,
    "SEED": 0n,
  },
  "metadata": {},
  "moniker": undefined,
  "nonce": 0n,
  "publicKey": "134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e",
  "stakes": {},
  "validator": {
    "address": "2370a54078de4a12becff8746e1c2826c9ef8b92",
    "power": 0n,
    "publicKey": "publicKey",
  },
}
`;

exports[`Repository > should index an account and find it by its validator address 1`] = `
{
  "address": "bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr",
  "balances": {
    "BEAR": 0n,
    "FIRE": 0n,
    "GOLD": 0n,
    "SEED": 0n,
  },
  "lockedBalances": {
    "BEAR": 0n,
    "FIRE": 0n,
    "GOLD": 0n,
    "SEED": 0n,
  },
  "metadata": {},
  "moniker": "moniker",
  "nonce": 0n,
  "publicKey": "134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e",
  "stakes": {},
  "validator": {
    "address": "2370a54078de4a12becff8746e1c2826c9ef8b92",
    "power": 0n,
    "publicKey": "publicKey",
  },
}
`;

exports[`Repository > should index an account and find it by its validator moniker 1`] = `
{
  "address": "bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr",
  "balances": {
    "BEAR": 0n,
    "FIRE": 0n,
    "GOLD": 0n,
    "SEED": 0n,
  },
  "lockedBalances": {
    "BEAR": 0n,
    "FIRE": 0n,
    "GOLD": 0n,
    "SEED": 0n,
  },
  "metadata": {},
  "moniker": "moniker",
  "nonce": 0n,
  "publicKey": "134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e",
  "stakes": {},
  "validator": {
    "address": "2370a54078de4a12becff8746e1c2826c9ef8b92",
    "power": 0n,
    "publicKey": "publicKey",
  },
}
`;
